hydra-main/src/main/java/com/addthis/hydra/job/store/CachedSpawnDataStore.java

```diff
@@ -0,0 +1,151 @@
+                                    key.getRight().length() :
+                                    0;
+
+                        // Multiply strlen by 2 (full width characters in java
+                        return 2 * (value.length() + leftWeight + rightWeight);
+                    }
+                })
+                .maximumWeight(dataStoreCacheSize)
+                .build(new CacheLoader<Pair<String, String>, String>() {
+                    @Override
+                    public String load(Pair<String, String> key) throws Exception {
+                        String path = key.getLeft();
+                        String childId = key.getRight();
+
+                        if (childId == null) {
+                            return CachedSpawnDataStore.this.dataStore.get(path);
+                        } else {
+                            return CachedSpawnDataStore.this.dataStore.getChild(path, childId);
+                        }
+                    }
+                });
+    }
+
+    @Override
+    public String getDescription() {
+        return dataStore.getDescription();
+    }
+
+    @Override
+    public String get(String path) {
+        try {
+            return cache.get(defaultKey(path));
+        } catch (ExecutionException | CacheLoader.InvalidCacheLoadException e) {
+            log.error("failed to execute get from cache", e);
+            return null;
+        }
+    }
+
+    @Override
+    public Map<String, String> get(String[] paths) {
+        List<String> notCached = new ArrayList<>();
+        Map<String, String> results = new TreeMap<>();
+
+        // Load every path from the cache that we can. For all the paths that do not exist in the cache, continue to
+        // do the batched query logic which should be implemented in the underlying DataStore's `get(String[])`
+        for (String path : paths) {
+            String result = cache.getIfPresent(defaultKey(path));
+            if (result == null) {
+                notCached.add(path);
+            } else {
+                results.put(path, result);
+            }
+        }
+
+        String[] remainingPaths = new String[notCached.size()];
+        remainingPaths = notCached.toArray(remainingPaths);
+
+        Map<String, String> resultsFromDB = dataStore.get(remainingPaths);
+
+        for (Map.Entry<String, String> entry : resultsFromDB.entrySet()) {
+            String path = entry.getKey();
+            cache.put(defaultKey(path), entry.getValue());
+        }
+
+        results.putAll(resultsFromDB);
+        return results;
+    }
+
+    @Override
+    public void put(String path, String value) throws Exception {
+        dataStore.put(path, value);
+        cache.put(defaultKey(path), value);
+    }
+
+    @Override
+    public void putAsChild(String parent, String childId, String value) throws Exception {
+        dataStore.putAsChild(parent, childId, value);
+        cache.put(ImmutablePair.of(parent, childId), value);
+    }
+
```

Comment:
What if something goes wrong with the data store put, or it say, just takes a rather long while, and meanwhile the cache entry gets evicted, possibly repopulated from the still-not-updated database, and then finally the database put succeeds.

With that ordering, it is possible to have all kinds of inconsistent values, switching back and forth (possibly permanently). I think the current ordering is safest.